### PR TITLE
Add capacity base random policy

### PR DESCRIPTION
**Pull request by dengweisysu**
used for cluster of workers with different capacity.
Close Alluxio/alluxio#15012

### CODE REVIEW

File: core/client/fs/src/main/java/alluxio/client/block/policy/CapacityBaseRandomPolicy.java
Comment (maobaolong) [inline core/client/fs/src/main/java/alluxio/client/block/policy/CapacityBaseRandomPolicy.java:23]: Remove this extra empty line as java.* and javax.* are in the same import group

**Review by maobaolong — CHANGES_REQUESTED** [review]
@dengweisysu Thanks for this policy contribution, before deep review, please fix the style issue, thank you.

File: core/client/fs/src/main/java/alluxio/client/block/policy/CapacityBaseRandomPolicy.java
Comment (beinan) [inline core/client/fs/src/main/java/alluxio/client/block/policy/CapacityBaseRandomPolicy.java:40]: Looks like it returns zero when the range of o1 equals or includes the range of o2,  but how about the range of o2 includes the range of o1?

File: core/client/fs/src/main/java/alluxio/client/block/policy/CapacityBaseRandomPolicy.java
Comment (beinan) [inline core/client/fs/src/main/java/alluxio/client/block/policy/CapacityBaseRandomPolicy.java:55]: nit: might be better if use a stream rather than a for loop.  your call.

File: core/client/fs/src/main/java/alluxio/client/block/policy/CapacityBaseRandomPolicy.java
Comment (beinan) [reply core/client/fs/src/main/java/alluxio/client/block/policy/CapacityBaseRandomPolicy.java:40]: I see, I think the range of o2 is just one, but you won't put 0 capacity worker into the list.  But it looks like too magic to me.
```
binarySearch(workerList, new CapacityRangeWorkerInfo(randomLong, randomLong + 1, null),
```

File: core/client/fs/src/main/java/alluxio/client/block/policy/CapacityBaseRandomPolicy.java
Comment (beinan) [inline core/client/fs/src/main/java/alluxio/client/block/policy/CapacityBaseRandomPolicy.java:70]: I might use a single value treemap to implement this algorithm rather than a range.  You could use the current capacity as the key and the worker index as the value.  Then you can use the lowerEntry(randomLong) to get the worker you want.

File: core/client/fs/src/test/java/alluxio/client/block/policy/CapacityBaseRandomPolicyTest.java
Comment (beinan) [inline core/client/fs/src/test/java/alluxio/client/block/policy/CapacityBaseRandomPolicyTest.java:41]: Shall we add a couple of negative test cases?  e.g. your getworker might return null

File: core/client/fs/src/main/java/alluxio/client/block/policy/CapacityBaseRandomPolicy.java
Comment (dengweisysu) [reply core/client/fs/src/main/java/alluxio/client/block/policy/CapacityBaseRandomPolicy.java]: Thanks for your good suggestion. I have done with much simplified implement.

File: core/client/fs/src/test/java/alluxio/client/block/policy/CapacityBaseRandomPolicyTest.java
Comment (dengweisysu) [reply core/client/fs/src/test/java/alluxio/client/block/policy/CapacityBaseRandomPolicyTest.java]: add one more test: testNoMatchWorker

File: core/client/fs/src/main/java/alluxio/client/block/policy/CapacityBaseRandomPolicy.java
Comment (dengweisysu) [reply core/client/fs/src/main/java/alluxio/client/block/policy/CapacityBaseRandomPolicy.java]: use forEach in new implement.

**Comment by dengweisysu** [comment]
wecom-temp-b17b7b03f9621f6cd04f8019e1a455bc
the fail tests seems nothing to do with my commit. 
@beinan look forward to your review

File: core/client/fs/src/main/java/alluxio/client/block/policy/CapacityBaseRandomPolicy.java
Comment (jiacheliu3) [reply core/client/fs/src/main/java/alluxio/client/block/policy/CapacityBaseRandomPolicy.java:43]: You may want to include this case in the test too, for example have multiple workers with same capacity and see what happens. This can actually be a very common case where majority of workers have the same capacity.

Comment (dengweisysu) [reply core/client/fs/src/main/java/alluxio/client/block/policy/CapacityBaseRandomPolicy.java:43]: > use the capacity as key, two workers with the same capacity (unlucky)

the key is the floor of the capacity range,  not its capacity.  the key is grow with getAndAdd method
eg. three workers with 10T capacity. the first key is 0 (range 0~10), the second key is 10 (range 10~20), the third key is 20 ( range 20~30)

Comment (dengweisysu) [reply core/client/fs/src/main/java/alluxio/client/block/policy/CapacityBaseRandomPolicy.java:43]: > You may want to include this case in the test too, for example have multiple workers with same capacity and see what happens. This can actually be a very common case where majority of workers have the same capacity.

@jiacheliu3 so, I think is not necessary to add more test for multiple workers with same capacity for my explanation above

**Comment by dengweisysu** [comment]
> 

done. I add these methods like MostAvailableFirstPolicy

**Comment by dengweisysu** [comment]
> 

done. I add these methods like MostAvailableFirstPolicy

**Comment by dengweisysu** [comment]
> @dengweisysu Thanks for the improvement! Left a comment on the algorithm. Also i see other BlockLocationPolicy implementations have `equals(), hashCode() and toString()`, can you add those too?
> 
> Rest of the code look good. I will add some minor comments after your next commit.

done. I add these methods like MostAvailableFirstPolicy

File: core/client/fs/src/test/java/alluxio/client/block/policy/CapacityBaseRandomPolicyTest.java
Comment (jiacheliu3) [inline core/client/fs/src/test/java/alluxio/client/block/policy/CapacityBaseRandomPolicyTest.java:32]: I suspect all these empty `WorkerNetAddress` equal to each other. Could you double check?

File: core/client/fs/src/main/java/alluxio/client/block/policy/CapacityBaseRandomPolicy.java
Comment (jiacheliu3) [reply core/client/fs/src/main/java/alluxio/client/block/policy/CapacityBaseRandomPolicy.java:54]: Good catch. I didn't notice the flooring.

I think it's still necessary to add some more tests for multiple workers with the same capacity. You can just duplicate your existing `getSuccessWorker()` and make them `getWorkerDifferentCapacity()` and `getWorkerSameCapacity()`.

File: core/client/fs/src/main/java/alluxio/client/block/policy/CapacityBaseRandomPolicy.java
Comment (jiacheliu3) [inline core/client/fs/src/main/java/alluxio/client/block/policy/CapacityBaseRandomPolicy.java:48]:
```suggestion
    // The key is the floor of capacity range the worker cover
    // For example if worker1 has 100B and worker2 has 100B
    // The map will have {0L -> worker1, 100L -> worker2}
```

File: core/client/fs/src/main/java/alluxio/client/block/policy/CapacityBaseRandomPolicy.java
Comment (jiacheliu3) [inline core/client/fs/src/main/java/alluxio/client/block/policy/CapacityBaseRandomPolicy.java:77]: These implementations are a bit strange, especially the hashCode returns 0 so this is purely dummy. However I get it that you are doing the same as `MostAvailableFirstPolicy`. Also I don't find any places that rely on the hashCode of `BlockLocationPolicy` so I think it's fine.

File: core/client/fs/src/main/java/alluxio/client/block/policy/CapacityBaseRandomPolicy.java
Comment (jiacheliu3) [inline core/client/fs/src/main/java/alluxio/client/block/policy/CapacityBaseRandomPolicy.java:42]: nit: no need for extra newline

File: core/client/fs/src/main/java/alluxio/client/block/policy/CapacityBaseRandomPolicy.java
Comment (jiacheliu3) [inline core/client/fs/src/main/java/alluxio/client/block/policy/CapacityBaseRandomPolicy.java:32]:
```suggestion
/**
 * Randomly distribute workload based on the worker capacities so bigger workers get more requests.
 * The randomness is based on the capacity instead of availability because in the long run,
 * all workers will be filled up and have availability close to 0.
 * We do not want the policy to degenerate to all workers having the same chance.
 */
```

**Review by jiacheliu3 — CHANGES_REQUESTED** [review]
Mostly looks good to me. The comments are mainly about tests and code comments.
Thanks for the improvements!

File: core/client/fs/src/test/java/alluxio/client/block/policy/CapacityBaseRandomPolicyTest.java
Comment (dengweisysu) [reply core/client/fs/src/test/java/alluxio/client/block/policy/CapacityBaseRandomPolicyTest.java:32]: thanks for your good suggestions. I done with assertSame instead.

**Comment by dengweisysu** [comment]
@jiacheliu3 I have just  applied your suggestion, please review again. Much appreciated.

File: core/client/fs/src/test/java/alluxio/client/block/policy/CapacityBaseRandomPolicyTest.java
Comment (jiacheliu3) [inline core/client/fs/src/test/java/alluxio/client/block/policy/CapacityBaseRandomPolicyTest.java:89]:
```suggestion
   * @param targetValue must be in [0,totalCapacity)
```

**Review by jiacheliu3 — APPROVED** [review]
Great work! LGTM with a minor comment in java comment. Good to merge once that's updated.

Also there's no need to force push every time. It's fine to create multiple commits in one PR, because when we merge one PR, all commits will be squashed into one. It's easier to track commits and see what's changed.

File: core/client/fs/src/test/java/alluxio/client/block/policy/CapacityBaseRandomPolicyTest.java
Comment (dengweisysu) [reply core/client/fs/src/test/java/alluxio/client/block/policy/CapacityBaseRandomPolicyTest.java]: @jiacheliu3 done

File: core/client/fs/src/main/java/alluxio/client/block/policy/CapacityBaseRandomPolicy.java
Comment (ggezer) [reply core/client/fs/src/main/java/alluxio/client/block/policy/CapacityBaseRandomPolicy.java:78]: Right, let's not supply `hashCode` and `equals` with dummies and can just drop those to revert behavior to java default.

Also, `toStringHelper` with no fields is also as good as not providing a `toString()` implementation at all.

File: core/client/fs/src/main/java/alluxio/client/block/policy/CapacityBaseRandomPolicy.java
Comment (ggezer) [inline core/client/fs/src/main/java/alluxio/client/block/policy/CapacityBaseRandomPolicy.java:59]: This is not quite right. The client policy's job is to pick the best worker based on the policy it implements.
IMO, total capacity being 0 should just revert this policy to a purely random selection among available workers. `null` should be returned when there are no workers to work with, not when there's not one best worker.

File: core/client/fs/src/main/java/alluxio/client/block/policy/CapacityBaseRandomPolicy.java
Comment (ggezer) [inline core/client/fs/src/main/java/alluxio/client/block/policy/CapacityBaseRandomPolicy.java:54]: nit: `capacityRangeStart` ?

File: core/client/fs/src/main/java/alluxio/client/block/policy/CapacityBaseRandomPolicy.java
Comment (ggezer) [inline core/client/fs/src/main/java/alluxio/client/block/policy/CapacityBaseRandomPolicy.java:49]: AFAIU, this just assigns rings to workers in the continuous interval based on their capacity.
If correct, could you please enhance this comment with the such information as well? (Ideally pictorial explanation) So that the reader has an idea of how this logic achieves capacity based randomness.

**Review by ggezer — CHANGES_REQUESTED** [review]
Thanks @dengweisysu for this capability.
Left some comments.

File: core/client/fs/src/main/java/alluxio/client/block/policy/CapacityBaseRandomPolicy.java
Comment (beinan) [reply core/client/fs/src/main/java/alluxio/client/block/policy/CapacityBaseRandomPolicy.java:49]: +1

File: core/client/fs/src/main/java/alluxio/client/block/policy/CapacityBaseRandomPolicy.java
Comment (maobaolong) [inline core/client/fs/src/main/java/alluxio/client/block/policy/CapacityBaseRandomPolicy.java:66]: why you have to override these three method?

File: core/client/fs/src/main/java/alluxio/client/block/policy/CapacityBaseRandomPolicy.java
Comment (jiacheliu3) [reply core/client/fs/src/main/java/alluxio/client/block/policy/CapacityBaseRandomPolicy.java:59]: Thx for the catch

File: core/client/fs/src/main/java/alluxio/client/block/policy/CapacityBaseRandomPolicy.java
Comment (jiacheliu3) [reply core/client/fs/src/main/java/alluxio/client/block/policy/CapacityBaseRandomPolicy.java:66]: It was originally my suggestion as I saw other `BlockLocationPolicy` implementations have overriden these. I wasn't sure if the `hashCode()` and `equals()` were used somewhere.

**Comment by dengweisysu** [comment]
@jiacheliu3 @ggezer 
totalCapacity is zero only when there is no worker has any capacity. it doesn't mean that there is no available workers?

**Comment by beinan** [comment]
> @jiacheliu3 @ggezer totalCapacity is zero only when there is no worker has any capacity. it doesn't mean that there is no available workers?

gotcha, I think we should return null in this case,  @jiacheliu3 & @ggezer what do you guys think?

**Comment by ggezer** [comment]
> > @jiacheliu3 @ggezer totalCapacity is zero only when there is no worker has any capacity. it doesn't mean that there is no available workers?
> 
> gotcha, I think we should return null in this case, @jiacheliu3 & @ggezer what do you guys think?

storage space is not mandated for workers. I'd return `null` only when given workerinfo list is empty. And fall back to fully random when there's no capacity to work with.

**Comment by beinan** [comment]
> > > @jiacheliu3 @ggezer totalCapacity is zero only when there is no worker has any capacity. it doesn't mean that there is no available workers?
> > 
> > 
> > gotcha, I think we should return null in this case, @jiacheliu3 & @ggezer what do you guys think?
> 
> storage space is not mandated for workers. I'd return `null` only when given workerinfo list is empty. And fall back to fully random when there's no capacity to work with.

I see, so there are two possibilities when we see the totalCapacity is zero.
1. the workerinfo list is empty -- in this case we should return null
2. the workerinfo list is not empty but none of workers in the list has any capacity, in this case, we should fallback to fully random.  
Thank you @ggezer for the clarification!  I'm convinced, and I think the fallback is necessary for this PR

**Comment by maobaolong** [comment]
@beinan @ggezer @jiacheliu3 So, we can extends `RoundRobinPolicy` for this new policy, right?

**Comment by jiacheliu3** [comment]
> > > > @jiacheliu3 @ggezer totalCapacity is zero only when there is no worker has any capacity. it doesn't mean that there is no available workers?
> > > 
> > > 
> > > gotcha, I think we should return null in this case, @jiacheliu3 & @ggezer what do you guys think?
> > 
> > 
> > storage space is not mandated for workers. I'd return `null` only when given workerinfo list is empty. And fall back to fully random when there's no capacity to work with.
> 
> I see, so there are two possibilities when we see the totalCapacity is zero.
> 
> 1. the workerinfo list is empty -- in this case we should return null
> 2. the workerinfo list is not empty but none of workers in the list has any capacity, in this case, we should fallback to fully random.
>    Thank you @ggezer for the clarification!  I'm convinced, and I think the fallback is necessary for this PR

If all workers have 0 capacity, why not fall back to LocalFirst? Since Alluxio is not keeping at cache at all, I think the best we can do is to provide as much locality as possible.

**Comment by maobaolong** [comment]
wecom-temp-1169be73b2c1b56fae132cf107f8b943
wecom-temp-48da6954927f2a80d89834edd9ed8daf

@jiacheliu3 @ggezer @beinan After a scan of other policies, i found that both RoundRollbin and LocalFirst are nothing to do with a zero capacity worker, so i guess we could return null `if totalCapacity == 0`, any thoughts?

**Comment by ggezer** [comment]
As per recent findings by @maobaolong, even though it's not the right thing to do, It's OK to conform to common behavior.

**Review by beinan — APPROVED** [review]
lgtm, many thanks to both @dengweisysu and @maobaolong !

File: core/client/fs/src/main/java/alluxio/client/block/policy/CapacityBaseRandomPolicy.java
Comment (jiacheliu3) [reply core/client/fs/src/main/java/alluxio/client/block/policy/CapacityBaseRandomPolicy.java:49]: I believe this will suffice:
```
// All the capacities will form a ring of continuous intervals
// And we throw a dice in the ring and decide which worker to pick
// For example if worker1 has capacity 10, worker2 has 20, worker3 has 40,
// the ring will look like [0, 10), [10, 30), [30, 70).
// A key in the map is the LHS of a range.
// So the map will look like {0 -> w1, 10 -> w2, 30 -> w3}.
```

**Review by jiacheliu3 — APPROVED** [review]
Thanks for the improvements! Just some minor comments to go and good to merge after those minor comments are resolved @maobaolong @dengweisysu

**Comment by dengweisysu** [comment]
@jiacheliu3 @ggezer new commit with comments enhanced, and rename two variables: capacityRangeStart and rangeStartMap

**Comment by jiacheliu3** [comment]
@dengweisysu i believe you missed some comments on removing the dummy implementation of `hashCode()` and `toString()` and `equals()`. Could you remove those too?

**Comment by dengweisysu** [comment]
@jiacheliu3 done

**Review by maobaolong — APPROVED** [review]
@dengweisysu Thanks for this useful policy contribution. LGTM.

Look forward to see the further contribution.

**Comment by jiacheliu3** [comment]
Failed tests are irrelevant to the change
```
Error: 1.356 [ERROR] alluxio.util.JvmPauseMonitorTest.testMockedInfoPause  Time elapsed: 0.716 s  <<< FAILURE!
java.lang.AssertionError
	at org.junit.Assert.fail(Assert.java:87)
	at org.junit.Assert.assertTrue(Assert.java:42)
	at org.junit.Assert.assertTrue(Assert.java:53)
	at alluxio.util.JvmPauseMonitorTest.testMockedInfoPause(JvmPauseMonitorTest.java:103)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.rules.ExpectedException$ExpectedExceptionStatement.evaluate(ExpectedException.java:258)
	at org.junit.rules.ExternalResource$1.evaluate(ExternalResource.java:54)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at org.junit.runners.BlockJUnit4ClassRunner$1.evaluate(BlockJUnit4ClassRunner.java:100)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:366)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:103)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:63)
	at org.junit.runners.ParentRunner$4.run(ParentRunner.java:331)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:79)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:329)
	at org.junit.runners.ParentRunner.access$100(ParentRunner.java:66)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:293)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:413)
	at org.apache.maven.surefire.junit4.JUnit4Provider.execute(JUnit4Provider.java:364)
	at org.apache.maven.surefire.junit4.JUnit4Provider.executeWithRerun(JUnit4Provider.java:289)
	at org.apache.maven.surefire.junit4.JUnit4Provider.executeTestSet(JUnit4Provider.java:237)
	at org.apache.maven.surefire.junit4.JUnit4Provider.invoke(JUnit4Provider.java:158)
	at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:428)
	at org.apache.maven.surefire.booter.ForkedBooter.execute(ForkedBooter.java:162)
	at org.apache.maven.surefire.booter.ForkedBooter.run(ForkedBooter.java:562)
	at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:548)
```

**Review by ggezer — APPROVED** [review]
Thanks @dengweisysu

**Comment by jiacheliu3** [comment]
alluxio-bot, merge this please

**Pull request merged**
